File: Easy/206 - ReverseLinkedList/Solution.js
/**
 * Question 206. Reverse Linked List (https://leetcode-cn.com/problems/reverse-linked-list/)
 *
 * 执行用时: 64 ms, 在所有 JavaScript 提交中击败了 88.95% 的用户
 * 内存消耗: 35.1 MB, 在所有 JavaScript 提交中击败了 78.16% 的用户
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let currentHead = null;

  while (head != null) {
    let currentNode = currentHead;

    currentHead = head;
    head = head.next;

    currentHead.next = currentNode;
  }

  return currentHead;
};

File: Easy/206 - ReverseLinkedList/Solution.cs
/// Question 206. Reverse Linked List (https://leetcode-cn.com/problems/reverse-linked-list/)

/**
 *
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution
{
    public ListNode ReverseList(ListNode head)
    {
        // 执行用时: 104 ms, 在所有 C# 提交中击败了 90.42% 的用户
        // 内存消耗: 24.5 MB, 在所有 C# 提交中击败了 7.14% 的用户
        ListNode reversedNode = null;

        while (head != null)
        {
            ListNode current = reversedNode;
            reversedNode = new ListNode(head.val);
            reversedNode.next = current;
            head = head.next;
        }

        return reversedNode;
    }

    public ListNode ReverseListWithoutNewNode(ListNode head)
    {
        // 执行用时: 108 ms, 在所有 C# 提交中击败了 74.65% 的用户
        // 内存消耗: 24.1 MB, 在所有 C# 提交中击败了 7.14% 的用户
        ListNode reversedNode = null;

        while (head != null)
        {
            ListNode current = reversedNode;

            reversedNode = head;
            head = head.next;

            reversedNode.next = current;
        }

        return reversedNode;
    }
}